src/kademlia/dht_storage.cpp

```diff
@@ -346,6 +345,7 @@ namespace
 		}
 
 		void put_immutable_item(sha1_hash const& target
+			, sha1_hash const& our_id
```

Comment:
Well, why do you think it doesn't make sense? Don't your reasoning in 1) implies a natural (almost-)partitioning of the incoming data?

Regarding the `dht_default_storage` implementation detail you mention, if this is precisely the "default implementation" why not put here the criteria to discard the elements? Why computing the minimum distance to any client's node ids is better, if that probably takes as an "effective" reference id, one that not properly represent the one from the "DHT network" they are coming(are requested)?